### ISSUE TITLE
accidentally typing `cabal (non-command) <path>` gives bogus output instead of reporting that 'non-command' does not exist

### ISSUE THREAD

**Issue by chessai** [haskell/cabal#5600]
minimal reproducible example:

```bash
$ git clone git@github.com:chessai/semirings && cd semirings
$ cabal new-sdist
$ cabal new-upload dist-newstyle/sdist/semirings-0.2.1.1.tar.gz
Resolving dependencies...
Build profile: -w ghc-8.6.1 -O1
In order, the following will be built (use -v for more details):
 - fake-package-0 (exe:script) (configuration changed)
Configuring executable 'script' for fake-package-0..
Preprocessing executable 'script' for fake-package-0..
Building executable 'script' for fake-package-0..

Main.hs:1:1: error: lexical error at character '\US'
  |
1 | �������}kw�Ȳ��̯�}��6�`;N�3�
  | ^

```

Here's a picture:

![what](https://user-images.githubusercontent.com/18648043/46310277-f0274980-c58c-11e8-9d36-1d7a91212611.png)

Compare the above to the following, which behaves correctly:

```bash
$ cabal new-upload dist-newstyle/sdist/semirings-0.2.1.1.tar.gz --publish
cabal: unrecognised command: new-upload (try --help)
```

**Comment by typedrat** [comment]
That's not specific to `new-upload` as a (non) command, it's because if there's not a command that it knows about but there is a path, it's jumping into the script interpreter logic for some reason.

**Comment by chessai** [comment]
The gibberish kind of looks like it's trying to parse the tar.gz file. If you run `less dist-newstyle/sidst/semirings-0.2.1.1.tar.gz`, you'll get similar-looking gibberish

**Comment by leonschoorl** [comment]
The same behavior can also be triggered by passing an unknown flag to an existing command.
And you don't even need a .cabal file.
```
$ mkdir tempdir
$ cd tempdir
$ touch some-file
$ cabal new-clean --foo some-file
Resolving dependencies...
Build profile: -w ghc-8.4.3 -O1
In order, the following will be built (use -v for more details):
 - fake-package-0 (exe:script) (first run)
Configuring executable 'script' for fake-package-0..
Preprocessing executable 'script' for fake-package-0..
Building executable 'script' for fake-package-0..
[1 of 1] Compiling Main             ( Main.hs, /tmp/cabal-repl.-7293/dist-newstyle/build/x86_64-linux/ghc-8.4.3/fake-package-0/x/script/build/script/script-tmp/Main.o )

Main.hs:1:1: error:
    Could not find module ‘Prelude’
    Use -v to see a list of the files searched for.
```

It seems this happens when there should be an error in parsing the arguments (unknown subcommand or unknown flag) and the last argument is a filename that exists.

```
$ cabal new-clean --foo non-existant-file
cabal: unrecognized 'new-clean' option `--foo'
$ cabal new-clean some-file --foo
cabal: unrecognized 'new-clean' option `--foo'
$ cabal new-clea non-existant-file
cabal: unrecognised command: new-clea (try --help)
$ cabal new-clea some-file
Resolving dependencies...
Build profile: -w ghc-8.4.3 -O1
In order, the following will be built (use -v for more details):
 - fake-package-0 (exe:script) (first run)
Configuring executable 'script' for fake-package-0..
...
```

**Comment by hvr** [comment]
For the record, in Git `master` this now manifests differently due to the improved `{-cabal`-block scanner (c6a11601142e4a986af0995f9168c0e41fab28ff): 

```
$ cabal new-clean --foo some-file
cabal: Failed extracting script block: `{- cabal:` start marker not found
```

**Comment by hvr** [comment]
I talked to @typedrat and I believe we'll see a PR fixing this quite soon :-)

**Comment by hvr** [comment]
Fyi, as this has been cherry-picked into the 2.4 branch, this will be fixed in a future cabal 2.4.x point-release (assuming there is one)

**Comment by chessai** [comment]
Thanks @typedrat @hvr
